### PR TITLE
290 feature introduce metadatafunctiontransformer for ingestion of metadata at inference time

### LINKED ISSUE #290: feature: Introduce MetadataFunctionTransformer for ingestion of metadata at inference time
**Is your feature request related to a problem? Please describe.**
Preprocessing is traditionally treated as a static operation tied to the training dataset. However, production is dynamic; instrument light source power, thermal drift, and batch-specific reference spectra directly influence the signal. Currently, integrating these real-time variables into a `scikit-learn.Pipeline` requires complex workarounds that often bypass the standard API. This forces scientists to choose between maintaining a clean, reproducible Pipeline and achieving scientific accuracy through adaptive compensation. This lack of a native "metadata-aware" path creates a barrier to deploying robust, production-ready spectral pipelines.

**Describe the solution you'd like**
I propose adding a `MetadataFunctionTransformer` to `chemotools.adaptation` that acts as a standardized API for metadata routing. By wrapping custom functions in this transformer, we provide a structured, production-ready interface for metadata-dependent preprocessing while preserving full scientific flexibility for the user's custom processing logic. 

**Describe alternatives you've considered**
- Standard `FunctionTransformer`: While efficient for simple transformations, it remains "metadata-blind," necessitating external orchestration that complicates the `Pipeline` object.

- Custom Estimator Classes: While flexible, building a custom class for every processing function introduces significant maintenance overhead and risks inconsistency in data validation and API compliance. S

**Additional context**
This enhancement will allow users to focus on the science—the functions themselves—rather than the plumbing of the routing protocol. I have developed a modular, production-ready prototype that separates the validation logic into a reusable utility, ensuring consistency across all future `chemotools` components.

Proactive Verification: Implementing an automated signature validation utility (via `inspect`) to ensure that the metadata provided by the user perfectly aligns with the processing function's requirements, catching configuration errors before data processing begins.

### PR DESCRIPTION
**Pull request by paucablop**
## Summary

- Add `MetadataFunctionTransformer`: a new transformer that wraps an arbitrary callable and routes named metadata through a scikit-learn `Pipeline` via the metadata-routing API.
- Add `chemotools.adaptation.functions`: a module of predefined, picklable functions for common spectral adaptation operations.
- Add `check_metadata_signature` validation utility to `chemotools._validation`.

## Why is this change needed?

- Spectral preprocessing pipelines often require auxiliary per-sample data (e.g. reference spectra, wavelength arrays) that must be passed alongside `X`. The standard sklearn API provides no clean way to thread such metadata through a `Pipeline` with a user-supplied callable. `MetadataFunctionTransformer` fills this gap.
- Lambdas and local functions are not picklable, which breaks `check_estimator`, `GridSearchCV`, and joblib parallelism. The predefined functions in `adaptation.functions` are module-level and fully picklable.

## Closes

- #290 

## Related issues

- 

## Type of change

- [ ] Bug fix
- [x] New feature
- [ ] Refactor
- [ ] Performance improvement
- [ ] Documentation update
- [ ] Tests only
- [ ] CI / DevOps / tooling
- [ ] Dependency update
- [ ] Breaking change

## What changed?

- Added `chemotools/adaptation/_metadata_function_transformer.py`: `MetadataFunctionTransformer` class with `fit`, `transform`, `fit_transform`, and `get_metadata_routing` methods; validates parameter constraints and checks that `metadata` keys match the callable's signature via `check_metadata_signature`.
- Added `chemotools/adaptation/functions.py`: four module-level picklable functions — `subtract_reference`, `divide_by_reference`, `scale_by_factor`, `add_offset`.
- Updated `chemotools/adaptation/__init__.py`: exports `MetadataFunctionTransformer` and the `functions` submodule.
- Updated `chemotools/_validation.py`: added `check_metadata_signature` — validates that the keys declared in `metadata` are present in the callable's signature and that required positional arguments are covered.

## What did NOT change?

- `XAxisInterpolator` and all other existing adaptation transformers are unchanged; their `set_fit_request`/`set_transform_request` pattern is intentionally preserved.
- No changes to public APIs of any existing class.

## API and compatibility impact

- [ ] No public API changes
- [x] Public API added
- [ ] Public API changed
- [ ] Deprecation introduced
- [ ] Breaking change introduced

### Notes
- scikit-learn API compatibility: `_parameter_constraints` uses bare `callable` (consistent with `sklearn.preprocessing.FunctionTransformer`); `get_metadata_routing()` override is required because metadata keys are dynamic (only known at construction time).
- Affected modules/classes/functions: `chemotools.adaptation.MetadataFunctionTransformer`, `chemotools.adaptation.functions`, `chemotools._validation.check_metadata_signature`.
- Backward compatibility considerations: fully additive; no existing behaviour changed.

## Validation

- [ ] `task format-check`
- [ ] `task lint`
- [ ] `task spelling`
- [ ] `task type-check`
- [x] `task test`
- [x] `task coverage`
- [ ] `task test:matrix`
- [ ] `task docs:check`
- [ ] `task build`

### Validation notes
- `task check` exits 0; all 18 new tests pass.
- `check_estimator` passes with a documented `expected_failed_checks` dict covering the 13 known sklearn failures that are inherent to any metadata-routing transformer (transform requires metadata to produce non-trivial output).

## Tests

- [x] Added new tests
- [ ] Updated existing tests
- [ ] No tests added because this PR only changes docs / CI / metadata

### Test coverage details
- Relevant test files: `tests/adaptation/test_metadata_function_transformer.py` (13 tests), `tests/adaptation/test_functions.py` (5 tests).
- Edge cases covered: non-callable `func` raises `ValueError`; metadata key not in function signature raises `ValueError`; required positional argument not declared in `metadata` raises `ValueError`; `**kwargs` functions are accepted without error; extra metadata keys passed at transform time are silently ignored; `fit_transform` threads metadata to both `fit` and `transform`; `get_metadata_routing` registers all declared keys for both `fit` and `transform`.
- Numerical / estimator behavior checked: all four predefined functions verified for numerical correctness; `check_estimator` compliance verified for `MetadataFunctionTransformer` with no metadata and with each of the four predefined functions.

## Documentation

- [ ] Docs updated
- [x] Docstrings updated
- [ ] No docs update needed

### Documentation notes
- Full NumPy-style docstrings on `MetadataFunctionTransformer` and all functions in `adaptation.functions`.
- Docstring example in `MetadataFunctionTransformer` is a runnable `doctest`.

## Dependency / build / CI impact

- [x] No dependency changes
- [ ] Runtime dependency change
- [ ] Dev dependency change
- [ ] GitHub Actions / CI changed
- [ ] Build/release process changed

### Notes
- No new imports outside the existing `sklearn`, `numpy`, and stdlib (`inspect`, `collections.abc`) dependencies.

## Reviewer focus

Please focus on:
- `get_metadata_routing()` implementation — confirm the `add_request(param=key, alias=True)` registration is correct for both `fit` and `transform` routing.
- `check_metadata_signature` in `_validation.py` — confirm the logic for detecting orphaned keys and missing required positional arguments is sound, especially the `**kwargs` pass-through case.

## Screenshots / logs / benchmark output

<details>
<summary>task check output</summary>
</details>

```text
task check → exit 0
18 new tests passing (13 in test_metadata_function_transformer.py, 5 in test_functions.py)
```
Checklist
- [x] I linked the relevant issue(s) or explained why none exists
- [x] I kept this PR scoped to a single purpose
- [x] I added or updated tests where appropriate
- [x] I updated documentation where appropriate
- [x] I verified the change locally using the relevant tasks above
- [x] I considered backward compatibility and public API impact
- [x] I am ready for review

### CODE REVIEW

File: chemotools/adaptation/_metadata_function_transformer.py [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+
+        Returns
+        -------
+        self : MetadataFunctionTransformer
+        """
+        # Validate the input parameters
+        self._validate_params()
+
+        # Validate the X data
+        validate_data(self, X, ensure_2d=True, reset=True, dtype="numeric")
```
[the remark this answers is not in the document]
Comment (paucablop) [reply chemotools/adaptation/_metadata_function_transformer.py]: Already implemented in commit cd0e947b06e5d92cefbc9392c40ea2feb25e50ef

[the remark this answers is not in the document]
Comment (Gnpd) [reply chemotools/adaptation/_metadata_function_transformer.py]: Is this solved?
Should we still need the `if self.validate:`?

File: chemotools/adaptation/_metadata_function_transformer.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+            ``self.metadata`` are forwarded.
+
+        Returns
+        -------
+        X_transformed : np.ndarray of shape (n_samples, n_features)
+            The result of calling ``func(X, **kwargs)``.
+        """
+
+        # Ensures .fit() was called by checking for trailing underscore attributes
+        check_is_fitted(self)
```
Comment (Gnpd) [inline chemotools/adaptation/_metadata_function_transformer.py]: It should be possible to call transform with an unfitted estimator, right?

Comment (paucablop) [reply chemotools/adaptation/_metadata_function_transformer.py]: I went back and forth with the idea of having this. I left it for consistency with the other transformers, but might not be needed (and when looking at the [FunctionTransformer](https://github.com/scikit-learn/scikit-learn/blob/cc50648cc1b759b53a4edbce0f3bb6c237349448/sklearn/preprocessing/_function_transformer.py#L237) it looks like they do not require fitting the transformer before calling `transform()`.  So maybe I could just remove this line!

File: chemotools/adaptation/functions.py [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+    --------
+    >>> import numpy as np
+    >>> from chemotools.adaptation.functions import divide_by_reference
+    >>> X = np.array([[2.0, 4.0, 6.0], [8.0, 10.0, 12.0]])
+    >>> ref = np.array([[2.0, 2.0, 2.0]])
+    >>> divide_by_reference(X, ref)
+    array([[1., 2., 3.],
+           [4., 5., 6.]])
+    """
+    return X / _check_metadata(reference, "reference")
```
Comment (Gnpd) [inline chemotools/adaptation/functions.py]: We can divide by 0 here if the data has 0s, right? Should we cover that?

**Review by Gnpd — DISMISSED** [review]
